### DISCUSSION TITLE
[Feature Request] Check if the cursor is moved away from the selection and moved back

### DISCUSSION THREAD

**Discussion by chengtie** [microsoft/monaco-editor#3987] [Q&A]
### Context

- [X] This issue is not a bug report. *(please use a different template for reporting a bug)*
- [X] This issue is not a duplicate of an existing issue. *(please use the [search](https://github.com/microsoft/monaco-editor/issues) to find existing issues)*

### Description

I have a web application by ReactJS and NodeJS. It contains a Monaco Editor. I register a hover provider by [registerHoverProvider](https://microsoft.github.io/monaco-editor/typedoc/functions/languages.registerHoverProvider.html). The use scenario is as follows:

1. the user selects a text
2. the user hovers over the selected text
3. a diagnostic window is open with `Loading...`, OpenAI API is launched with the selected text (and some contextual information), and returns the result in several seconds, we then display it in the diagnostic window.

Now, we realize that it happens often that the user moves its hover over the selected text. Every time he moves the cursor, the `provideHover` function is triggered. As a consequence, several OpenAI requests with same parameters are launched in a very short time, which is a waste.

**So I would like to know if there is a conventional way to avoid this?**

One option can be to use cache. We can save OpenAI requests and their parameters in a cache. Then, before launching a request, we check if it exists in the cache. If yes, we return the previous result; otherwise we launch the request. And we need to clear the cache from time to time. I would like to especially clear the cache when the user moves the cursor away from the selection area and moves back to the selection.

Does anyone know how to check **if the user moves the cursor away from the selection and moves back to the selection**?

### Monaco Editor Playground Link

[Link](https://microsoft.github.io/monaco-editor/playground.html?source=v0.38.0#XQAAAAJbBgAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscw_UAhCnGBGpWlpVktocO7_BJCkQTDFxt_ci1cqyNnm0w49r1zMV-o8vPWY8dB3Tj3dFwfIUYjYEcjgDIeqku-ycAqJKQWBvWy8nur-rHf2d7mpmAFwBkgbF7n7fKAdlnTlCDmPA443Mq-KLXal95AabCnj7aukYulXSchQcMnbI-qyB_FOioTblMejQtKjD5YkngRUzC1Jpcrc2vzOT6AqakYgZCxAfREn2wzvFsyKM8HeqRK2G7UW7Z4SgGcEyccMGLGFn8mS50dpd_extTHCexQW2475Q6Vr7vbaeGWP1DN7tEqj5S1LQYEw9jUf4ceTpJ1kP6BX3zUU0p2tiVd51NV5cpYYrXZsWcpWDgCvPveBe7-zXFWHVbFy3qC7NlzCKEI2PGjnlsr-M5iFGctwqtydLE4zJkrTx2dRDp1xDXkMYkwndk-qTfz4lLgsl4JOBEdhYJLNvLQ63hhmU5HD69dITiA-axiNTWDjzvKWy3OVpsZQ8dlGjCYuauC9ixysrAD6tu4rG26m-Re7EhXOaswgrT_ySC_qMSdkUsQWFiZPF6P_tpGRRh7UdlXyRFhKf4xonr62qrp-AVDVqkND8473G919kAjLk_aLXcNc-JznH9UX1q0oAjpWfBFQ--DOaMeSxjkWQqBx56a2fcnXVDYA1QtOi37X8d-uoHgmHJtVBFyvDWGeIx3csEioimCiRSJsrWHHEFx0j_rrsUG5hNKVpefhjhClK4MgFwplpnXhQgO7CSLmkJNtc-k2BBancRLdkRYq342GGDDlngfH162pX0wN4E1hlFvllKVaP2A43bCKue1XncThuuJz4CkyvUoTWssarsV0SomHp1qC4v7J79xNLFU1_-q-PX7)

In this code, I use `simulatedApiCall` to simulate an OpenAI request. We can select a text, hover on it. We can notice in the console that the API requests can be called again when we move the cursor in the selection area.

### Monaco Editor Playground Code

```typescript
monaco.languages.register({ id: "mySpecialLanguage" });

function simulatedApiCall(text) {
    console.log("launch the request");
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            const result = `API request result for "${text}"`; // Simulated result
            resolve(result);
        }, 3000); // Delay for 3 seconds to simulate network latency
    });
}

let myEditor = null; // Store the editor instance here

myEditor = monaco.editor.create(document.getElementById("container"), {
    value: "\n\nHover over this text",
    language: "mySpecialLanguage",
});

monaco.languages.registerHoverProvider("mySpecialLanguage", {
    provideHover: function (model, position) {
        const selectedText = myEditor.getModel().getValueInRange(myEditor.getSelection());

        // If nothing is selected, don't do anything
        if (!selectedText) {
            return null;
        }
        
        return simulatedApiCall(selectedText)
            .then(res => {
                console.log("res", res)
                return {
                    range: new monaco.Range(
                        1,
                        1,
                        model.getLineCount(),
                        model.getLineMaxColumn(model.getLineCount())
                    ),
                    contents: [{ value: res }],
                };
            })
            .catch(error => {
                console.error(error)
                // Return null in case of an error
                return null;
            });
    },
});
```

**Comment by hediet** [comment]
This is not supported by the API. Maybe you can try to directly use the internal services.

We also don't plan to support this scenario. Feel free to create a feature request for more concrete API, ideally describing how we would implement that and why this is usefull for others too.

**Reply by chengtie** [reply]
Hello, what are "internal services"?
